File: android/src/main/java/com/reactnativenitrofreeze/RNNitroFreezeModule.java
package com.reactnativenitrofreeze;

import android.view.View;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.UiThreadUtil;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.uimanager.NativeViewHierarchyManager;
import com.facebook.react.uimanager.UIBlock;
import com.facebook.react.uimanager.UIManagerModule;

/**
 * Native module for react-native-nitro-freeze.
 * Provides native-level freeze optimizations for Android views.
 */
@ReactModule(name = RNNitroFreezeModule.NAME)
public class RNNitroFreezeModule extends ReactContextBaseJavaModule {
    public static final String NAME = "RNNitroFreeze";

    public RNNitroFreezeModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    /**
     * Set freeze state on a native view.
     * 
     * When frozen (frozen=true):
     * - Set view to INVISIBLE (removed from draw but keeps layout)
     * - Disable drawing optimizations (setWillNotDraw)
     * - Disable clickable and focusable
     * 
     * When unfrozen (frozen=false):
     * - Restore VISIBLE state
     * - Re-enable drawing
     * - Restore interactive properties
     * 
     * @param viewTag React Native view tag
     * @param frozen true to freeze, false to unfreeze
     */
    @ReactMethod
    public void setViewFrozen(final int viewTag, final boolean frozen) {
        ReactApplicationContext context = getReactApplicationContext();
        UIManagerModule uiManager = context.getNativeModule(UIManagerModule.class);
        
        if (uiManager != null) {
            uiManager.addUIBlock(new UIBlock() {
                @Override
                public void execute(NativeViewHierarchyManager nativeViewHierarchyManager) {
                    final View view = nativeViewHierarchyManager.resolveView(viewTag);
                    
                    if (view == null) {
                        return;
                    }
                    
                    UiThreadUtil.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            if (frozen) {
                                // Freeze the view
                                view.setVisibility(View.INVISIBLE);
                                view.setWillNotDraw(true);
                                view.setClickable(false);
                                view.setFocusable(false);
                                view.setEnabled(false);
                            } else {
                                // Unfreeze the view
                                view.setVisibility(View.VISIBLE);
                                view.setWillNotDraw(false);
                                view.setClickable(true);
                                view.setFocusable(true);
                                view.setEnabled(true);
                            }
                        }
                    });
                }
            });
        }
    }
}

File: android/src/main/java/com/margelo/nitrofreeze/RNNitroFreezeModule.kt
package com.margelo.nitrofreeze

import android.view.View
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.common.UIManagerType

/**
 * RNNitroFreezeModule
 * 
 * Native Android module for optimizing frozen views.
 * Implements view freezing at the Android View level for maximum performance.
 */
class RNNitroFreezeModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "RNNitroFreeze"

    /**
     * Set the frozen state of a view.
     * 
     * When frozen:
     * - Sets visibility to INVISIBLE
     * - Disables drawing (willNotDraw = true)
     * - Disables touch events
     * - Pauses animations
     * 
     * When unfrozen:
     * - Restores visibility to VISIBLE
     * - Enables drawing (willNotDraw = false)
     * - Enables touch events
     * - Resumes animations
     * 
     * @param viewTag The React Native view tag
     * @param frozen Whether to freeze or unfreeze the view
     */
    @ReactMethod
    fun setViewFrozen(viewTag: Int, frozen: Boolean) {
        UiThreadUtil.runOnUiThread {
            val view = findView(viewTag) ?: return@runOnUiThread
            
            if (frozen) {
                freezeView(view)
            } else {
                unfreezeView(view)
            }
        }
    }

    /**
     * Find a view by its React tag.
     * Works with both Fabric and legacy renderer.
     */
    private fun findView(viewTag: Int): View? {
        return try {
            // Try Fabric first (new architecture)
            val fabricUIManager = UIManagerHelper.getUIManager(
                reactApplicationContext,
                UIManagerType.FABRIC
            )
            fabricUIManager?.resolveView(viewTag)
        } catch (e: Exception) {
            try {
                // Fall back to legacy renderer
                val legacyUIManager = UIManagerHelper.getUIManager(
                    reactApplicationContext,
                    UIManagerType.DEFAULT
                )
                legacyUIManager?.resolveView(viewTag)
            } catch (e: Exception) {
                null
            }
        }
    }

    /**
     * Apply freeze optimizations to a view
     */
    private fun freezeView(view: View) {
        // Make invisible (but keep in layout, unlike GONE)
        view.visibility = View.INVISIBLE
        
        // Disable drawing
        view.setWillNotDraw(true)
        
        // Disable touch events
        view.isEnabled = false
        view.isClickable = false
        view.isFocusable = false
        
        // Pause animations
        view.animate().cancel()
        view.clearAnimation()
        
        // Tag the view so we know it's frozen
        view.setTag(R.id.freeze_state, true)
    }

    /**
     * Remove freeze optimizations from a view
     */
    private fun unfreezeView(view: View) {
        // Restore visibility
        view.visibility = View.VISIBLE
        
        // Enable drawing
        view.setWillNotDraw(false)
        
        // Enable touch events
        view.isEnabled = true
        
        // Clear freeze tag
        view.setTag(R.id.freeze_state, false)
    }

    companion object {
        const val NAME = "RNNitroFreeze"
    }
}

